Race condition in unit tests using cmpopts.EquateApproxTime

### What happened?

I ran `make test` and curiously got an error without even modifying the code. I looked deeper into it and it looks like tests that use `cmpopts.EquateApproxTime` will fail (sometimes) if executed repeatedly.
The two affected packages are `internal/controller/apiextensions/composite` and `internal/controller/apiextensions/claim`.

### How can we reproduce it?

```
go test -count=1000 ./internal/controller/apiextensions/composite
go test -count=1000 ./internal/controller/apiextensions/claim
```

<details><summary>Test failure from ./internal/controller/apiextensions/composite</summary>

```nocode
--- FAIL: TestReconcile (0.00s)
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot compose resources: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot publish connection details: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Available"),
                                                        "status":             string("True"),
                                                        "type":               string("Ready"),
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot fetch Composition: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot configure composite resource: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                ... // 2 identical entries
                                "metadata": map[string]any{"annotations": map[string]any{"crossplane.io/paused": string("")}},
                                "spec":     map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Available"),
                                                        "status":             string("True"),
                                                        "type":               string("Ready"),
                                                },
                                        },
        -                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:26Z")},
        +                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:27Z")},
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "metadata":   map[string]any{"annotations": map[string]any{"crossplane.io/paused": string("true")}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcilePaused"),
                                                        "status":             string("False"),
                                                        "type":               string("Synced"),
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "metadata":   map[string]any{"deletionTimestamp": string("2023-04-21T09:46:26Z")},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Deleting"),
                                                        "status":             string("False"),
                                                        "type":               string("Ready"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot unpublish connection details: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "metadata":   map[string]any{"deletionTimestamp": string("2023-04-21T09:46:26Z")},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Deleting"),
                                                        "status":             string("False"),
                                                        "type":               string("Ready"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("refusing to use invalid Composition: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot select CompositionUpdatePolicy: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Creating"),
                                                        "status":             string("False"),
                                                        "type":               string("Ready"),
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot select Composition: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Available"),
                                                        "status":             string("True"),
                                                        "type":               string("Ready"),
                                                },
                                        },
        -                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:26Z")},
        +                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:27Z")},
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "spec":       map[string]any{"compositionRef": map[string]any{}},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("ReconcileSuccess"),
                                                        "status":             string("True"),
                                                        "type":               string("Synced"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Available"),
                                                        "status":             string("True"),
                                                        "type":               string("Ready"),
                                                },
                                        },
        -                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:26Z")},
        +                               "connectionDetails": map[string]any{"lastPublishedTime": string("2023-04-21T09:46:27Z")},
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "metadata":   map[string]any{"deletionTimestamp": string("2023-04-21T09:46:26Z")},
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "reason":             string("Deleting"),
                                                        "status":             string("False"),
                                                        "type":               string("Ready"),
                                                },
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot remove composite resource finalizer: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
    reconciler_test.go:780: WantComposite(...): -want, +got:   &composite.Unstructured{
                Unstructured: unstructured.Unstructured{
                        Object: map[string]any{
                                "apiVersion": string(""),
                                "kind":       string(""),
                                "status": map[string]any{
                                        "conditions": []any{
                                                map[string]any{
        -                                               "lastTransitionTime": string("2023-04-21T09:46:26Z"),
        +                                               "lastTransitionTime": string("2023-04-21T09:46:27Z"),
                                                        "message":            string("cannot add composite resource finalizer: boom"),
                                                        "reason":             string("ReconcileError"),
                                                        ... // 2 identical entries
                                                },
                                        },
                                },
                        },
                },
          }
FAIL
FAIL    github.com/crossplane/crossplane/internal/controller/apiextensions/composite    0.035s
FAIL
```

</details>

--- a/internal/controller/apiextensions/claim/reconciler_test.go
+++ b/internal/controller/apiextensions/claim/reconciler_test.go
@@ -831,7 +831,7 @@ func TestReconcile(t *testing.T) {
 					return s
 				}
 				return ts
-			}), cmpopts.EquateApproxTime(30*time.Second)); diff != "" {
+			}), cmpopts.EquateApproxTime(3*time.Second)); diff != "" {
 				t.Errorf("\n%s\nr.Reconcile(...): -want, +got:\n%s", tc.reason, diff)
 			}
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {

--- a/internal/controller/apiextensions/claim/reconciler_test.go
+++ b/internal/controller/apiextensions/claim/reconciler_test.go
@@ -825,7 +825,13 @@ func TestReconcile(t *testing.T) {
 
 			got, err := r.Reconcile(context.Background(), reconcile.Request{})
 
-			if diff := cmp.Diff(tc.want.claim, tc.args.claim, cmpopts.EquateApproxTime(30*time.Second)); diff != "" {
+			if diff := cmp.Diff(tc.want.claim, tc.args.claim, cmpopts.AcyclicTransformer("StringToTime", func(s string) any {
+				ts, err := time.Parse(time.RFC3339, s)
+				if err != nil {
+					return s
+				}
+				return ts
+			}), cmpopts.EquateApproxTime(30*time.Second)); diff != "" {
 				t.Errorf("\n%s\nr.Reconcile(...): -want, +got:\n%s", tc.reason, diff)
 			}
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {

--- a/internal/controller/apiextensions/composite/reconciler_test.go
+++ b/internal/controller/apiextensions/composite/reconciler_test.go
@@ -775,7 +775,13 @@ func WantComposite(t *testing.T, want resource.Composite) func(ctx context.Conte
 		// Normally we use a custom Equal method on conditions to ignore the
 		// lastTransitionTime, but we may be using unstructured types here where
 		// the conditions are just a map[string]any.
-		diff := cmp.Diff(want, got, cmpopts.EquateApproxTime(3*time.Second))
+		diff := cmp.Diff(want, got, cmpopts.AcyclicTransformer("StringToTime", func(s string) any {
+			ts, err := time.Parse(time.RFC3339, s)
+			if err != nil {
+				return s
+			}
+			return ts
+		}), cmpopts.EquateApproxTime(3*time.Second))
 		if diff != "" {
 			t.Errorf("WantComposite(...): -want, +got: %s", diff)
 		}

